### ISSUE TITLE
[Outreachy 2024] Applicants tasks/contributions

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#31460]
# Intro

This ticket will serve as the central point tracking the contributions of perspective Outreachy (spring 2024 cohort) applicants.

For those interested, this process is described here: https://www.outreachy.org/apply/make-contributions/
The systemd project will participate again this year: https://www.outreachy.org/outreachy-june-2024-internship-round/communities/systemd/#add-linux-services-live-debugging-functionality-to

A series of tasks will be listed here. Some of them will be "repeatable", ie: some basic tasks that every applicant can do. Some can only be done once, so they are first-come first-served. Some will be relatively quick and should be doable in a couple hours, others might take several days, and they'll be tagged as such.

Every applicant should start with the basic repeatable tasks, and complete at least two. Simply quote the task in a comment, posting proof of completion. We will "thumbs up" the comment to acknowledge completion.

After at least two of those are done and acked, then applicants should move on to the "non-repeatable" tasks. Write a comment quoting the task you selected, and if it's open-ended specify what exactly you are going to do, before you start. Those should end up with a pull request. Note that simply opening a pull request is not enough, they will be reviewed and might need follow-ups before they can be merged, so take that into account when planning for time.

Applicants should complete and wait for ack on completion at least 2 of the repeatable tasks before moving on to the short/long tasks.

# Documentation for new contributors
Contributors who are new to `systemd` are highly encouraged to start from the following documents:

- https://systemd.io/CONTRIBUTING/
- https://systemd.io/CODING_STYLE/
- https://systemd.io/HACKING/
- https://systemd.io/CODE_QUALITY/
- https://systemd.io/ARCHITECTURE/

# Repeatable tasks

- clone the project, build it, add an `Hello World from <name>!` log message to a tool of your choice from `src/`, run it and post a screenshot as proof
- clone the project, build it, pick one tool under `src/` of your choice and run it, inspect the code and in your own word write a paragraph explaining what the code (or a section of the code, some are quite large!) does - note that we don't mean just from the point of view of a user, as that's explained in the manual, but describe the actual code
- study the documentation, and write a paragraph summarizing one of the documents in your own words. Pick at least one medium/large size document, OR at least three smaller ones. For easier viewing, documents under `docs/` are rendered on https://systemd.io/ and documents under `man/` are rendered on https://www.freedesktop.org/software/systemd/man/

# Short tasks
There is no specific order in which these tasks have to be picked, choose any.

- fix one of the issues labelled as "please-submit-as-pr" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr
- improve an existing unit test `src/test/` to increase code coverage for the basic common code - see under `src/basic/` on https://coveralls.io/github/systemd/systemd
- add a new unit test to cover the 'shared' common code - see under `src/shared/` on https://coveralls.io/github/systemd/systemd

# Long tasks
Please note that the description might sound simple, but in reality the task might require more time than you have available. Before starting, state in a comment which issue/TODO you'd like to tackle and how much time you have available, and we'll try to advise on whether it's enough to complete it. There is no specific order in which these tasks have to be picked, choose any.

- ~~Add a set of assertion macros to tests.h (ASSERT_OK(), ASSERT_EQ(), ASSERT_GE(), ...) that log the failed condition before crashing and convert a unit test file to use them~~
- fix one of the issues labelled as "RFE" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3A%22RFE+%F0%9F%8E%81%22
- fix one of the TODO items https://github.com/systemd/systemd/blob/main/TODO
- add an integration test `test/README.testsuite` for a tool/daemon that is not currently covered

**Comment by Unique-Usman** [comment]
Hello @bluca Thank you very much for this well detailed information. I was able to build the project on my system. 

I do not seem quite to understand well what you meant by adding `Hello World from <name>!` log message to a tool under src/. There are multiple directories inside src which mostly contain C code and header files.

**Comment by Unique-Usman** [comment]
![Screenshot from 2024-03-05 13-13-34](https://github.com/systemd/systemd/assets/86585626/5c154ca4-0547-4aee-b0d1-2f7dc26f1e0d)
Output
![Screenshot from 2024-03-05 13-16-59](https://github.com/systemd/systemd/assets/86585626/dba6abb4-b7c0-45a3-ab1c-1bd090966faa)
Input

This is for the repeatable task one.

**Comment by bluca** [comment]
Looks good, you may proceed with the next short repeatable task

**Comment by Unique-Usman** [comment]
The next short repeatable task which is mainly the combination of the last two 
1.
![Screenshot from 2024-03-06 01-57-31](https://github.com/systemd/systemd/assets/86585626/60be3dbc-8e13-48ec-b165-f48253f9b316)
2.
![Screenshot from 2024-03-06 01-57-31](https://github.com/systemd/systemd/assets/86585626/757c24e0-5caf-4c48-ad21-12ab89dc68f3)

**Comment by Unique-Usman** [comment]
![Screenshot from 2024-03-06 02-02-41](https://github.com/systemd/systemd/assets/86585626/82b5cf28-25de-48f5-b916-6dbd45bfb3c2)
Referenced code portion

**Comment by bluca** [comment]
> The next short repeatable task which is mainly the combination of the last two

Looks good, you may proceed with the next step, short tasks

**Comment by Chand-ra** [comment]
This is my submission for the repeatable task 1:

change:
![systemd_change](https://github.com/systemd/systemd/assets/114725427/ef432e95-92b8-4d36-8e3c-1c3665dd502c)

output:
![systemd_change_output](https://github.com/systemd/systemd/assets/114725427/ae16965b-9359-4996-a939-0596e2abc707)

**Comment by bluca** [comment]
> This is my submission for the repeatable task 1:

Looks good, you may proceed to the next step

**Comment by bluca** [comment]
> hello @bluca I am interested in this issue #29609. I have been able to see the file to be changed (70-mouse.hwdb). I also tried to find if there is any test for the hwdb.d but, there is none. Also, kindly give any suggestion regarding that.

hwdb is just a list of IDs, and requires the exact involved hardware to test, so I'd recommend to avoid it as a task

**Comment by Unique-Usman** [comment]
Thank you for that.

**Comment by bluca** [comment]
Sure, go for it. Our current style is "right justify", so please use that option.

**Comment by Unique-Usman** [comment]
@bluca should I proceed to take another task from "fix one of the issues labelled as "please-submit-as-pr" or proceed to "improving the existing unit test" ?

**Comment by bluca** [comment]
You can pick from any of the "short tasks", there's no order requirement. Please consider choosing something that involves code changes, as that makes it easier to move to the next stage

**Comment by Chand-ra** [comment]
This is my submission for repeatable task 2:

# ac-power
The systemd tool ‘ac-power’ detects whether the machine is connected to an external power source or not. It can be used as such:

![usage](https://github.com/systemd/systemd/assets/114725427/5d50a7aa-2c43-4e49-a162-6f718df8bdbb)
 
The code for the tool consists of a single file ‘ac-power.c’ and can be found in systemd/src/ac-power. It consists of three functions: **help()**, **parse_argv()** and **run()**. 

## help():
![help](https://github.com/systemd/systemd/assets/114725427/97d6b66b-f270-44ff-8c56-c31f900928ad)
 
The help() function is the simplest one and prints a string to stdout upon invocation. The string mentions the role of the tool and the possible arguments that can be passed.
## parse_argv():
![parse_argv1](https://github.com/systemd/systemd/assets/114725427/2450d498-bd8f-4e9e-969e-e0c405007a83)
![parse_argv2](https://github.com/systemd/systemd/assets/114725427/88226442-ff13-4058-99ad-50f8d53f7745)

This function parses the arguments passed to the tool through the help of GNU’s `getopt.h`. The start of the function defines an array of option structs which represents the valid paramters that can be passed to the tool. The struct consists of four fields:
- The first field represents the name of the option, it is a string. 
- The second field is an integer and tells us whether the option takes an argument. In this case, none of the options do. 
- The third and fouth fields assign a unique value to each option, through which that option can be identified. 

This array of structs can be then passed to the `getopt_long()` function which is responsible for actually decoding the options. The next few lines of assert() check for valid input and a variable (int c) is defined to store the return value of `getopt_long()`. 

Next, a while loop is run until the value of c is ‘-1’, which represents no more options being left. On each iteration of the loop, the updated value of ‘c’ is passed to a switch statement which determines the action needed to be performed. 

- The case c = ‘h’ calls the help() function which results in the help message being printed to stdout. 

- The case c = ‘v’ turns the value of the flag / global variable ‘arg_verbose’ to true, which is then parsed by run() to print ‘yes’ or ‘no’ to stdout, depending upon the options being passed. 

- The case c = ARG_VERSION means the option ‘--version’ was passed and hence, version() is called which prints the package version to stdout. 

- The case c = ARG_LOW represents the option ‘–-low’ being passed which simply changes the value of the enum variable arg_action to ACTION_LOW. The futher action that is required is performed inside of run().

- The value c = ‘?’ represents an invalid option being passed. If this case in encountered, `getopt_long()` has already printed an error message so the -EINVAL error code is simply returned. 

- The default case should never be reached so in the case that it is reached, the program is terminated with `assert_not_reached()`.
## run():
![run](https://github.com/systemd/systemd/assets/114725427/587553f0-2dbd-47e6-ab73-9fe35a60f990)
 
This function is defined as the entry point for the tool/program with the macro `DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE()`. The function starts by defining a variable (int r) to store the value to be returned. The next two functions calls, `log_parse_environment()` and `log_open()` prepare for logging, decoding environment variables and other such general functionality. Then, the `parse_argv()` function is called and its return value is stored in ‘r’. A value of 0 indicates the ‘—help’ or ‘--version’ option was passed and hence, the required message has already been printed and a negative value indicates an error. The program is terminated in both the cases.

Next, the arg_action flag is queried. Note that this flag has been assigned the required value through `parse_argv()`. If arg_action has value ‘ACTION_AC_POWER’, ‘r’ is updated with the return value of `on_ac_power()`, which is ‘1’ if ac power is connected, ‘0’ if not and ‘-1’ otherwise. An error is logged in case of r = -1.
If the value of arg_action has been assigned to ‘ACTION_LOW’ instead, the function `battery_is_discharging_or_low()` is called which, similar to the above case,  returns ‘1’ if the battery is discharging or low, ‘0’ if not and ‘-1’ otherwise. An error is logged in case of r = -1.

If the ‘arg_verbose’ flag has been set to true, ‘yes’ or ‘no’ is printed to stdout, depending upon the parameters being passed. At the end, the value opposite to that of ‘r’ is returned (since the function must return with positive failure), which is meant to be used by scripts and other functions.

**Comment by bluca** [comment]
> This is my submission for repeatable task 2:

Looks good, you may proceed with the next step

**Comment by devdekunle** [comment]
My submission for repeatable task1
Input Code:
![Screenshot from 2024-03-06 16-59-10](https://github.com/systemd/systemd/assets/110066922/23cd0057-9b67-4435-a6c6-92382ec27e88)
Output:
![Screenshot from 2024-03-06 16-59-10](https://github.com/systemd/systemd/assets/110066922/aa7d0603-77a3-4291-81a2-ee7275060b5d)

**Comment by devdekunle** [comment]
My Submission for repeatable Task 1
Input Code:
![Screenshot from 2024-03-06 16-59-23](https://github.com/systemd/systemd/assets/110066922/ddc1e8ba-be21-4349-8a80-8f0c07b98caf)
Output:
![Screenshot from 2024-03-06 16-59-10](https://github.com/systemd/systemd/assets/110066922/167bb80d-6e02-4cb4-98e0-b23f9c8dcde7)

**Comment by Chand-ra** [comment]
> Looks good, you may proceed with the next step

Right, I have a question regarding the next step. Do we absolutely need to work on an idea from the **short tasks** list or is it fine if we decide to work on a topic mentioned in the **long tasks** list as the next step?

**Comment by bluca** [comment]
> > Looks good, you may proceed with the next step
> 
> Right, I have a question regarding the next step. Do we absolutely need to work on an idea from the **short tasks** list or is it fine if we decide to work on a topic mentioned in the **long tasks** list as the next step?

Please start with a short task first

**Comment by bluca** [comment]
> My Submission for repeatable Task 1

Looks good, please go ahead

**Comment by devdekunle** [comment]
My submission for repeatable Task2
### **systemd-ask-password**
The systemd-ask-password utility provides a secure way for receiving a  password
![Screenshot from 2024-03-06 19-06-33](https://github.com/systemd/systemd/assets/110066922/e9145058-45dc-4d18-bfd4-849285c112a6)

The code which is located in the src/ask-password/ask-password.c file contains included standard C libraries for error handling, and other system-level operations.
It also contains three functions which include help(), parse_argv() and run() for successfully implementing this utility.
### The help() function:
![Screenshot from 2024-03-06 19-07-58](https://github.com/systemd/systemd/assets/110066922/29e6e5cf-31b9-45cf-958c-0bacfe5f6b21)
The help function defines the message displayed when the user runs systemd-ask-password command with the -h or --help option. The help function return 0 to indicate successful operation

### The parse_argv() function:
![Screenshot from 2024-03-06 19-08-05](https://github.com/systemd/systemd/assets/110066922/30377b14-9b73-4b0e-bd4f-bcdda492d495)
This function is responsible for parsing the command line arguments.
It defines constants for each command line long option using enumerations. it then defines an array of struct options representing the long options accepted by the program.
It then declares variables **c** and **r** for options parsing and resets `optind` to 0.
it sets and processes command line options and sets the value to c using  **`getopt_long`**.
using the values gotten from c from `getopt_long`, the switch statement handles each and sets corresponding variables
An if statement handles the --emoji options. it determines if to show an emoji based on the user's inputs.
it then sets the variable **arg_message** based on the remaining command line options or defaults to "input" if using echo. The function returns 1 on successful operation

### The run() function:
![Screenshot from 2024-03-06 19-08-41](https://github.com/systemd/systemd/assets/110066922/b89fcc40-b789-46af-b774-c0720df5e738)

The run function defines the main function for executing the utility. it the initializes logging, parses command lines arguements using the **parse_argv** function and sets its value to **r** and it also checks for errors. it sets a timeout for the password request and initializes a structure AskPasswordRequest with the information gotten from the command line arguments
It then invokes the **ask_password_auto** function to query the user for a password and sets the value to **r** .
It then prints the obtained password to stdout based on the options provided.he function returns 0 to indicate success.
lastly the module defines the main function using the DEFINE_MAIN_FUNCTION macro.

**Comment by bluca** [comment]
> My submission for repeatable Task2

Looks good, please go ahead with the next step

**Comment by Unique-Usman** [comment]
@bluca I added more unit test to cover uid_range_covers function in uid-range.c. But, after trying to test the whole project, I am getting this error. Kindly help to look at it.
![Screenshot from 2024-03-07 01-33-42](https://github.com/systemd/systemd/assets/86585626/a26a7fa0-ac22-4e0b-9c38-75cdb7c12d9b)
Thank you.

**Comment by bluca** [comment]
Run the test manually by simply calling the binary, and prefix it with SYSTEMD_LOG_LEVEL=debug. Also copy and paste text, not screenshots.

**Comment by Unique-Usman** [comment]
Thank you very much, I am sorry for the screenshots. I will copy and paste next time. The fail is not from my changes. I already pushed, kindly review https://github.com/systemd/systemd/pull/31666. Thank you very much.

**Comment by MarkTanui** [comment]
Hi @bluca 
I took time to learn how to set up well, and I'm now running everything well.
Here's my submission to repeatable task 1:

Added log message to systemd-notify

![image](https://github.com/systemd/systemd/assets/84702057/56a9a941-d5ae-4a20-a58f-33ad6f0a8e53)

Run & Output of tool

![image](https://github.com/systemd/systemd/assets/84702057/28ccec58-6294-40db-b1a9-cabc575844eb)

**Comment by Unique-Usman** [comment]
@bluca While going through the `src/test/` directory, I saw some files which do not have test. I specificaly take note of two of them
1. -> https://github.com/systemd/systemd/blob/main/src/basic/dirent-util.c 
2. -> https://github.com/systemd/systemd/blob/main/src/basic/login-util.c

Do they not require test file ? if not, I am thinking of trying to add a test file to it, kindly give any suggestion on how I can go about it. Thank you.

**Comment by bluca** [comment]
Yes having more tests is always good, you can look at existing test sources to see how they are integrated in the meson build system

**Comment by devdekunle** [comment]
@bluca while going through the issues page https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr I realized issue #30836 had been resolved and merged here #31069. But the issue is still visible and hasn't been closed. Could there be something no done yet?

**Comment by bluca** [comment]
> @bluca while going through the issues page https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr I realized issue #30836 had been resolved and merged here #31069. But the issue is still visible and hasn't been closed. Could there be something not done yet?

It was done, but just not linked so it wasn't automatically closed, it's been fixed now

**Comment by devdekunle** [comment]
@bluca I added unittestcases to expand coverage for the **argv_looks_like_help function** (located in the src/basic/argv-util.c module) to the **test-argv-utils.c** file. Please kindly review #31680.

**Comment by Chand-ra** [comment]
@bluca #31682 is my submission for the short tasks. Kindly let me know of any feedback whenever possible.

**Comment by devdekunle** [comment]
> # Intro
> This ticket will serve as the central point tracking the contributions of perspective Outreachy (spring 2024 cohort) applicants.
> 
> For those interested, this process is described here: https://www.outreachy.org/apply/make-contributions/ The systemd project will participate again this year: https://www.outreachy.org/outreachy-june-2024-internship-round/communities/systemd/#add-linux-services-live-debugging-functionality-to
> 
> A series of tasks will be listed here. Some of them will be "repeatable", ie: some basic tasks that every applicant can do. Some can only be done once, so they are first-come first-served. Some will be relatively quick and should be doable in a couple hours, others might take several days, and they'll be tagged as such.
> 
> Every applicant should start with the basic repeatable tasks, and complete at least two. Simply quote the task in a comment, posting proof of completion. We will "thumbs up" the comment to acknowledge completion.
> 
> After at least two of those are done and acked, then applicants should move on to the "non-repeatable" tasks. Write a comment quoting the task you selected, and if it's open-ended specify what exactly you are going to do, before you start. Those should end up with a pull request. Note that simply opening a pull request is not enough, they will be reviewed and might need follow-ups before they can be merged, so take that into account when planning for time.
> 
> Applicants should complete and wait for ack on completion at least 2 of the repeatable tasks before moving on to the short/long tasks.
> 
> # Documentation for new contributors
> Contributors who are new to `systemd` are highly encouraged to start from the following documents:
> 
> * https://systemd.io/CONTRIBUTING/
> * https://systemd.io/CODING_STYLE/
> * https://systemd.io/HACKING/
> * https://systemd.io/CODE_QUALITY/
> * https://systemd.io/ARCHITECTURE/
> 
> # Repeatable tasks
> * clone the project, build it, add an `Hello World from <name>!` log message to a tool of your choice from `src/`, run it and post a screenshot as proof
> * clone the project, build it, pick one tool under `src/` of your choice and run it, inspect the code and in your own word write a paragraph explaining what the code (or a section of the code, some are quite large!) does - note that we don't mean just from the point of view of a user, as that's explained in the manual, but describe the actual code
> * study the documentation, and write a paragraph summarizing one of the documents in your own words. Pick at least one medium/large size document, OR at least three smaller ones. For easier viewing, documents under `docs/` are rendered on https://systemd.io/ and documents under `man/` are rendered on https://www.freedesktop.org/software/systemd/man/
> 
> # Short tasks
> * fix one of the issues labelled as "please-submit-as-pr" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr
> * improve an existing unit test `src/test/` to increase code coverage for the basic common code - see under `src/basic/` on https://coveralls.io/github/systemd/systemd
> * add a new unit test to cover the 'shared' common code - see under `src/shared/` on https://coveralls.io/github/systemd/systemd
> 
> # Long tasks
> Please note that the description might sound simple, but in reality the task might require more time than you have available. Before starting, state in a comment which issue/TODO you'd like to tackle and how much time you have available, and we'll try to advise on whether it's enough to complete it.
> 
> * Add a set of assertion macros to tests.h (ASSERT_OK(), ASSERT_EQ(), ASSERT_GE(), ...) that log the failed condition before crashing and convert a unit test file to use them
> * fix one of the issues labelled as "RFE" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3A%22RFE+%F0%9F%8E%81%22
> * fix one of the TODO items https://github.com/systemd/systemd/blob/main/TODO
> * add an integration test `test/README.testsuite` for a tool/daemon that is not currently covered

According to the instructions, @bluca, for the short tasks, we are to expand **existing** unit tests for the src/basic/ directory and write **new** tests for the src/shared directory. 
Please are we strictly to write new tests for files without tests only from the src/shared directory or we can also write new tests for those files in src/basiic without tests?

**Comment by bluca** [comment]
Sure, you can add a new test for src/basic modules too

**Comment by MarkTanui** [comment]
@bluca  Submission of my Repeatable Tasks 2 & 3 combined.

Tool: `systemd-notify`

### Task 2
`systemd-notify`  allows services to communicate with `systemd` to send status updates or notifications. It is particularly useful for signaling the readiness of services or reporting errors during startup, which helps `systemd` manage dependencies between services more effectively.

`systemd-notify` as in `src/notfiy.c` has 5 functions:
1. `int help(void)` 
- Displays the help menu with the options available and a short description of each option. It also specifies the manual page using a `terminal_urlify_man` call.
![image](https://github.com/systemd/systemd/assets/84702057/e747c1fd-c7d9-4c35-844a-29bf276bcee9)

2.` pid_t manager_pid(void)`
- This function retrieves the PID of the service manager from the MANAGERPID environment variable. If the variable is not set or contains an invalid PID, it returns 0; otherwise, it returns the PID.
![image](https://github.com/systemd/systemd/assets/84702057/a04c7d48-dba7-4442-89c3-8f7ebcfca3a5)

3. `pid_t pid_parent_if_possible(void)`
- The function that returns the parent process ID if it's possible to obtain it. It checks if the parent process ID is less than or equal to 1 or if it's equal to the manager's process ID, and if so, it returns the current process ID. Otherwise, it returns the parent process ID. A PID less than or equal to 1 implies the process is either the kernel's scheduler process (PID 0) or the system's initial user-space process (PID 1), which oversees all other processes. Ordinary user or system processes started during the course of the system's operation will receive PIDs greater than 1.
![image](https://github.com/systemd/systemd/assets/84702057/804481f4-a288-4b6d-b491-d9d3160d2b19)

4. `int parse_argv(int argc, char *argv[])`
- The function parses command-line arguments using the `getopt_long` function. It uses a static array of option structures to define the long options and their corresponding short options. Based on the parsed options, it sets various variables and flags to control program behavior, and eventually returns a value to indicate success or failure.

- This function does the bulk of the work to ensure that arguments/options are well parsed and validated to know exactly which functions will be called. It uses the enum and struct data types together with switchcases and conditions to perfectly filter out the supplied arguments.

5. `int run(int argc, char* argv[])`
- The function parses command-line arguments to determine the action requested. Based on those actions, prepare a set of environment variables and/or messages. It also can optionally change the process's user or group ID.
Sends these notifications to systemd, and potentially waiting for acknowledgment if the processes are successfully tracked. After running, it cleans up before exiting, ensuring no unintended automatic notifications are sent.

### Task 3: Summarizing the `systemd-notify` man page and some of it's other related tool (`sd_notify()`).

#### Understanding `sd_notify()` Notifications
`systemd` uses a mechanism (sd_notify()) that allows services or programs that run in the background to inform `systemd` about their status. However, for `systemd` to correctly recognize which service a status update is coming from, certain conditions need to be met:

- When a background service sends a notification to `systemd` and then immediately quits, `systemd` may not be able to figure out which service sent the message. This is because `systemd` tries to match the notification with an active process. If the process has already exited, `systemd` cannot make this match and will ignore the message.

- If `systemd` started the process itself or if the process's permissions are set up in a specific way (using NotifyAccess=main or NotifyAccess=exec), systemd can track it more effectively. In these cases, even if the process sends a notification and exits, `systemd` knows which service it belonged to because it has been keeping track of it from the beginning.

Example: Imagine a service that starts up, performs a quick task, and then sends a "task complete" notification using `sd_notify()` before immediately exiting. If `systemd` hasn't been set up to track this process correctly, it won't know which service sent the "task complete" message and will ignore it.

#### How `systemd-notify` Works
`systemd-notify` is a tool that services can use to send these notifications more easily. It tries to make sure `systemd` correctly attributes the notification to the right service.

Initially, `systemd-notify` tries to send the notification as if it was sent by its parent process (the process that called `systemd-notify`). This makes sense when a script uses `systemd-notify` because it means the notification appears to come directly from the script rather than from `systemd-notify`. However, this requires certain privileges, and if `systemd-notify` doesn't have them, this step will fail.

Falling Back to Its Own Process ID - If the above doesn't work, `systemd-notify` will then send the notification using its own process ID.

The --pid= option can be used to control this behavior, allowing you to specify exactly which process ID (`systemd-notify` should pretend to be.

LMK if I can proceed to the short tasks.

**Comment by bluca** [comment]
> LMK if I can proceed to the short tasks.

Looks good, you can proceed

**Comment by MarkTanui** [comment]
@bluca, for my short tasks(fix one of the issues labeled as "please-submit-as-pr"), I decided to work on #20625

### Summary of Issue
The creator suggests a feature for `systemd` to allow central logging to the host namespace for units running in containers (lxc) with their own ` systemd-journald`.

**The creator of the issue attached a patch-set as POC. I have dug into the set to understand the proposed patches and how to merge them into a PR. However, before attempting to connect the patches, I wanted to highlight the patches and my understanding of them and ask for your guidance on how I could best proceed.**

Here are the summaries of each proposed patch and some code to be added/modified. Please LMK

###  Patch 1 Summary: 
 Introduces `HOSTONLY_JOURNALD` definition in (src/basic/log.c).  The definition is used to determine a couple of additional headers and checks. i.e.,` virt.h`, and two variables and conditions to check if the `systemd-journald` process is running in containers. If it is in a container, the `prohibit_ipc` flag is ignored with the expectation that connection is established to host namespace journals.
 
### Patch 2 Summary: 
This patch mainly focus on modifying (/src/journal/journald-server.c) to add unit attribute to the container unit logs. This is logical because in someway the logs from container units have to be identified as from a source.

As such, the changes add the unit name artificially (determined from the unit's group). if only, the `HOSTONLY_JOURNALD` is defined! It does so by adding the following condition:

changes:
```
+#ifdef HOSTONLY_JOURNALD
+                if (c->unit == NULL) {
+                    if (c->cgroup != NULL && strstr(c->cgroup, "/lxc.payload") != NULL) {
+                        char *name = strrchr(c->cgroup, '/');
+                        if (name) {
+                            IOVEC_ADD_STRING_FIELD(iovec, n, name+1, "_SYSTEMD_UNIT");
+                        }
+                    }
+                } else {
+#endif
+                    IOVEC_ADD_STRING_FIELD(iovec, n, c->unit, "_SYSTEMD_UNIT"); /* Unit names are bounded by UNIT_NAME_MAX */
+#ifdef HOSTONLY_JOURNALD
+                }
+#endif
```

### Patch 3 Summary: 
Modifies libsystemd's `sd_id128_get_boot` function to return the host boot_id. The patch modifies the sd_id128_get_boot function to try to use the host's boot_id across containers. This would make all journal entries, regardless of whether they're generated on the host or within containers, appear as if they come from the same boot session when viewed from the host's perspective.

The patch introduces a conditional compilation block with `#ifdef HOSTONLY_JOURNALD`,  which suggests that this code path is only included in builds where a single systemd-journald instance is meant to handle logs for both the host and containers.

It attempts to read the boot_id from a new location (HOST_BOOTID_PATH),  which is presumably a path to where the host's boot_id is made available (bind mounted) within the container's filesystem.

If reading from HOST_BOOTID_PATH fails (either because the file does not exist or is not accessible), it falls back to the existing method of reading the boot_id directly from the container's /proc/sys/kernel/random/boot_id.

changes:
```
+#ifdef HOSTONLY_JOURNALD
+                r = id128_read(HOST_BOOTID_PATH, ID128_UUID, &saved_boot_id);
+                if (r < 0) {
+#endif
+                        r = id128_read("/proc/sys/kernel/random/boot_id", ID128_UUID, &saved_boot_id);
+                        if (r < 0)
+                                return r;
+#ifdef HOSTONLY_JOURNALD
+                }
+#endif
```

### Patch 4 Summary: 

The patch defines a new configuration option, HOST_BOOTID_PATH, specifying the filesystem path at which the host's boot_id is made available to containers. This path is presumably used by containers to access the host's boot_id, ensuring that logs from containers are tagged with the host's boot session identifier, thus aligning with the changes discussed in the previous patch.

Changes:
```
+conf.set_quoted('HOST_BOOTID_PATH',                           'run/systemd/journal/host_boot_id')
+conf.set('HOSTONLY_JOURNALD',                                 1)
```

**Comment by devdekunle** [comment]
Good day @bluca kindly review my latest task #31689.

**Comment by bluca** [comment]
> for my short tasks(fix one of the issues labeled as "please-submit-as-pr"), I decided to work on #20625

Please avoid that, it was marked by mistake

**Comment by Tundroid** [comment]
Greetings @bluca , here is my first repeatable task.
Tool: **backlight**

![Source code:](https://github.com/systemd/systemd/assets/16786572/aad55db8-ebbe-4613-8697-772260a1f72f)

Output:
![Ourput](https://github.com/systemd/systemd/assets/16786572/7adf09be-5fb1-4a5a-a50a-35eee07bcf65)

**Comment by bluca** [comment]
> Greetings @bluca , here is my first repeatable task.

Looks good, please go ahead with the next step

**Comment by amehclibton** [comment]
Hello sir @bluca, this is my Repeatable Task 1:
![Uploading Screenshot 2024-03-13 at 13.58.08.png…]()
![Uploading Screenshot 2024-03-13 at 13.57.18.png…]()

**Comment by amehclibton** [comment]
Hello @bluca sir. This is my repeatable task 1:
<img width="1440" alt="Screenshot 2024-03-13 at 13 57 18" src="https://github.com/systemd/systemd/assets/162214367/62f8f08c-4527-46df-8532-b5a6c9a2fb66">
<img width="1440" alt="Screenshot 2024-03-13 at 13 58 08" src="https://github.com/systemd/systemd/assets/162214367/eaf25841-be06-4ccf-8d89-555f392de76b">

**Comment by Unique-Usman** [comment]
Hello @bluca  can I start with the long task after the merge of [?](https://github.com/systemd/systemd/pull/31700) If yes, I want to start with the first one, what is the target number of macros that we want to add? We have three named above, I guess ASSERT_LE() is also another one that we can add. 

I am thinking of having something similar to this. 

#define ASSERT_OK(expr) \
    do { 
        if (!(expr)) { 
            fprintf(stderr, "Assertion failed: %s\n", #expr); 
            abort(); 
        } 
    } while (false)

Is this okay ? 

Also are we doing it for all the tests on the repository ? 
Thank you. If there is also any other information that can help, kindly share. 
Thank you very much once again.

**Comment by bluca** [comment]
Yes, you can start on that. Use logging and braced groups. The details can be discussed in the PR. It is enough to start with one test file being converted.

**Comment by devdekunle** [comment]
Good evening @bluca. After my merges, Please for the long tasks, are supposed to do it in the order in which they appear?
if no, i want to do an issue in the todo list which involves making chase() return O_DIRECTORY if the path has a trailing slash. and also refusing resolution if the input has a trailing slash but the final node isn't a directory.

**Comment by bluca** [comment]
No ordering constraints, you can pick any that makes sense

**Comment by devdekunle** [comment]
okay thank you

**Comment by amehclibton** [comment]
Hello sir @bluca, this is my repeatable task 2.
![image](https://github.com/systemd/systemd/assets/162214367/5bccee48-3024-484c-940d-c611a2e1c054)

**Comment by amehclibton** [comment]
Hello sir @bluca my repeatable task 3: networkctl

networkctl is a tool used for the querying and modification of network status. To use it, you type the name of the tool, networkctl then a command. You could place options before the command or a network link after the command. The following commands are valid with this tool: list, status, lldp (link layer discovery protocol), label, delete, up, down, renew, forcerenew, reconfigure, reload, edit, cat. For example, using the status command will shows information about specific links. Along with the associated options, this tool assists administrators to manage and troubleshoot network interfaces within system.

**Comment by amehclibton** [comment]
Hello sir @bluca, Please let me know if I can begin on short tasks

**Comment by bluca** [comment]
> Hello sir @bluca, this is my repeatable task 2.Repeatable Task 2: ac-power.c

ac-power was already picked in https://github.com/systemd/systemd/issues/31460#issuecomment-1981074082 please choose another component that hasn't been done already

**Comment by amehclibton** [comment]
> > Hello sir @bluca, this is my repeatable task 2.Repeatable Task 2: ac-power.c
> 
> ac-power was already picked in [#31460 (comment)](https://github.com/systemd/systemd/issues/31460#issuecomment-1981074082) please choose another component that hasn't been done already

ok sir

**Comment by Unique-Usman** [comment]
Hello @bluca while waiting for https://github.com/systemd/systemd/pull/31853 to get resolved, can I start looking into RFE issue to work on or what is your suggestion, thank you.

**Comment by bluca** [comment]
> Hello @bluca while waiting for #31853 to get resolved, can I start looking into RFE issue to work on or what is your suggestion, thank you.

Yes, you can go ahead

**Comment by Unique-Usman** [comment]
@bluca please, can I work on this RFE https://github.com/systemd/systemd/issues/31743 ?  thank you.

**Comment by DaanDeMeyer** [comment]
@Unique-Usman I've already opened a PR to fix #31743 myself. I've linked the PR to the RFE now to make that clearer.

**Comment by Unique-Usman** [comment]
> @Unique-Usman I've already opened a PR to fix #31743 myself. I've linked the PR to the RFE now to make that clearer.

Noted. Thank you.

**Comment by Unique-Usman** [comment]
@bluca can I work on this https://github.com/systemd/systemd/issues/31856, thank you.

**Comment by bluca** [comment]
> @bluca can I work on this #31856, thank you.

yes, go ahead

**Comment by amehclibton** [comment]
> > Hello sir @bluca, this is my repeatable task 2.Repeatable Task 2: ac-power.c
> 
> ac-power was already picked in [#31460 (comment)](https://github.com/systemd/systemd/issues/31460#issuecomment-1981074082) please choose another component that hasn't been done already

Sir @bluca  this is my repeatable task 2:  networkd.c

This tool is responsible for managing network configurations and interfaces.

A.	Headers
1.	System Headers
-	netinet/in.h: Provides definition for IP addresses.
-	sys/stat.h:  Makes it possible to query files and directories
-	sys/types.h: Defines types used for system programming task such as file I/O management etc.
2.	Custom Headers
-	sd-daemon.h: Provides communication between system services and systemd
-	sd-event.h: Creates and manages loop events within applications 
-	bus-log-control-api.h: This header file defines an API related to logging control over D-Bus
-	capability-util.h: Contains definitions for Linux capabilities which has to do with control of privileges of a process. 
-	daemon-util: Contains definition for functions involved with background processes.
-	firewall-util.h: Contains declarations involved with firewall management functionalities.
-	main-func.h: Involved with assigning the main function in a program
-	mkdir-label.h: Contains definitions and functions to create directories with labels.
-	networkd-conf.h: Provides network configuration and management functionalities.
-	networkd-manager-bus.h: Involves management of network interfaces via the Dbus.
-	networkd-manager.h: Involved with managing network configurations and interfaces.
-	service-util.h: Involved with utilities for managing services associated with systemd.
-	signal-util.h: Involved with signal related functions and utilities.
-	strv.h: Contains macros and function definitions for working with string vectors.
-	user-util.h: Involves with user and group management functionalities.

B.	Functions
1.	static int run: 
-	log_setup(): Initialised in order to keep record of subsequent info which needs to be logged.
-	Service_parse_argv: Parses the command-line arguments received from the program
-	umask(0022): ensures that the files and directories created later on have the 0644 and 0755 permissions respectively.
-	if (argc != 0): To make sure there’s morethan 1 argument to proceed else prints that the program does not have any arguments
-	if (geteuid() == 0): This condition statement comes in to checks if the program is running as root. If yes, it drops privilages.
-	FOREACH_STRING(): Creates new files and directories needed during operations
-	r = manager_new(&m, /* test_mode = */ false): Creates a manager
-	Then next on the manager is initialized and made to perform various tasks and return error logs incase a task fails

C.	Other statements:
-	DEFINE_MAIN_FUNCTION(run): Defines run to be the main function here.

**Comment by devdekunle** [comment]
Hello @bluca, I found two tasks in the TODO list that I would be interested in.
1. introduce an option (or replacement) for "systemctl show" that outputs all
  properties as JSON, similar to busctl's new JSON output. In contrast to that
  it should skip the variant type string though.
 2. maybe add a tool that displays most recent journal logs as QR code to scan
  off screen and run it automatically on boot failures, emergency logs and
  such. Use DRM APIs directly, see
  https://github.com/dvdhrm/docs/blob/master/drm-howto/modeset.c for an example
  for doing that.
  
  Please I would like to know your take on them with regards to existing implementations and maybe time expectancy for them.
  Though as regards imolementation, I checked the systemctl show --output=json and It outputs key-value pairs, same as just systemctl show. Which i feel could mean it hasn't been implemented. So please can i get your take on the two?

**Comment by bluca** [comment]
The second one is already done, you can work on the first one

**Comment by bluca** [comment]
> > > Hello sir @bluca, this is my repeatable task 2.Repeatable Task 2: ac-power.c
> > 
> > 
> > ac-power was already picked in [#31460 (comment)](https://github.com/systemd/systemd/issues/31460#issuecomment-1981074082) please choose another component that hasn't been done already
> 
> Sir @bluca this is my repeatable task 2: networkd.c
> 
> This tool is responsible for managing network configurations and interfaces.
> 
> A. Headers
> 
> 1. System Headers
> 
> * netinet/in.h: Provides definition for IP addresses.
> * sys/stat.h:  Makes it possible to query files and directories
> * sys/types.h: Defines types used for system programming task such as file I/O management etc.
> 
> 2. Custom Headers
> 
> * sd-daemon.h: Provides communication between system services and systemd
> * sd-event.h: Creates and manages loop events within applications
> * bus-log-control-api.h: This header file defines an API related to logging control over D-Bus
> * capability-util.h: Contains definitions for Linux capabilities which has to do with control of privileges of a process.
> * daemon-util: Contains definition for functions involved with background processes.
> * firewall-util.h: Contains declarations involved with firewall management functionalities.
> * main-func.h: Involved with assigning the main function in a program
> * mkdir-label.h: Contains definitions and functions to create directories with labels.
> * networkd-conf.h: Provides network configuration and management functionalities.
> * networkd-manager-bus.h: Involves management of network interfaces via the Dbus.
> * networkd-manager.h: Involved with managing network configurations and interfaces.
> * service-util.h: Involved with utilities for managing services associated with systemd.
> * signal-util.h: Involved with signal related functions and utilities.
> * strv.h: Contains macros and function definitions for working with string vectors.
> * user-util.h: Involves with user and group management functionalities.
> 
> B. Functions
> 
> 1. static int run:
> 
> * log_setup(): Initialised in order to keep record of subsequent info which needs to be logged.
> * Service_parse_argv: Parses the command-line arguments received from the program
> * umask(0022): ensures that the files and directories created later on have the 0644 and 0755 permissions respectively.
> * if (argc != 0): To make sure there’s morethan 1 argument to proceed else prints that the program does not have any arguments
> * if (geteuid() == 0): This condition statement comes in to checks if the program is running as root. If yes, it drops privilages.
> * FOREACH_STRING(): Creates new files and directories needed during operations
> * r = manager_new(&m, /* test_mode = */ false): Creates a manager
> * Then next on the manager is initialized and made to perform various tasks and return error logs incase a task fails
> 
> C. Other statements:
> 
> * DEFINE_MAIN_FUNCTION(run): Defines run to be the main function here.

Looks good, you can proceed

**Comment by devdekunle** [comment]
> The second one is already done, you can work on the first one

Okay thanks

**Comment by amehclibton** [comment]
Sir @bluca I would love to try out issue #29666

**Comment by bluca** [comment]
> Sir @bluca I would love to try out issue #29666

Do you have that hardware?

**Comment by amehclibton** [comment]
> > Sir @bluca I would love to try out issue #29666
> 
> Do you have that hardware?

No sir. Sir which Issue can I work on then? Apparently most of the other issues with require me to have their respective hardwares

**Comment by bluca** [comment]
There are other types of "short tasks", they do not have to be done in order

**Comment by amehclibton** [comment]
Sir what about this #31034 ?

**Comment by bluca** [comment]
That's not a good one either, again check the "short task" paragraph, you can work on adding/enhancing a unit test

**Comment by amehclibton** [comment]
> That's not a good one either, again check the "short task" paragraph, you can work on adding/enhancing a unit test

ok sir

**Comment by Unique-Usman** [comment]
Hi @bluca , how are you doing ?

I am currently facing some problem setting up the mkosi, though I was able to set it up, but, the bootctl seems not to work. Should I check out another RFE that does not deal with bootctl or what is your advice regarding that ? thank you?

Also, I was trying to write a test for this file https://github.com/systemd/systemd/blob/main/src/basic/sort-util.c

I have something like this, 

`result = xbsearch_r(&key, array, nmemb, sizeof(int), (comparison_userdata_fn_t)cmp_int, NULL);`

But, there was a warning when I was trying to build the project for testing. the `comparison_userdata_fn_t` takes in three argument while the `cmp_int` and `cmpt_uint16` takes in two argument. But, there is actually a function `comparison_fn_t` which takes in two argument. I am thinking of changing the `comparison_userdata_fn_t` to  `comparison_fn_t` . I also I could not find where the function x_bsearch was used. Kindly give advice on this. Thank you.

**Comment by Unique-Usman** [comment]
Hello @bluca, some question which I have regarding the integration test.
1. I got to know that this link https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/control#L121 contains the depencies to run the integration test, my question is do we have to install of the depency ?
2. When I  checked the coverall, I saw some  percentage and some of them seems to be for unit test, how do I know which one is the unit test and which one is the integration test? Also, I do not really understand the structure in the coverall very well. 

Thank you very much, I will be expecting your reply.

**Comment by Unique-Usman** [comment]
https://github.com/systemd/systemd/issues/31895 
https://github.com/systemd/systemd/issues/31824

I found these two potential issue. @bluca kindly help check if it something I can work on. Thank you.

**Comment by bluca** [comment]
> Hello @bluca, some question which I have regarding the integration test.
> 
> 1. I got to know that this link https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/control#L121 contains the depencies to run the integration test, my question is do we have to install of the dependency ?
> 2. When I  checked the coverall, I saw some  percentage and some of them seems to be for unit test, how do I know which one is the unit test and which one is the integration test? Also, I do not really understand the structure in the coverall very well.
> 
> Thank you very much, I will be expecting your reply.

1) Yes
2) That is not tracked, it's a single coverage value for everything

**Comment by bluca** [comment]
> #31895 #31824
> 
> I found these two potential issue. @bluca kindly help check if it something I can work on. Thank you.

Avoid the first one, but you can do the second one

**Comment by Unique-Usman** [comment]
> Hi @bluca , how are you doing ?
> 
> I am currently facing some problem setting up the mkosi, though I was able to set it up, but, the bootctl seems not to work. Should I check out another RFE that does not deal with bootctl or what is your advice regarding that ? thank you?
> 
> Also, I was trying to write a test for this file https://github.com/systemd/systemd/blob/main/src/basic/sort-util.c
> 
> I have something like this,
> 
> `result = xbsearch_r(&key, array, nmemb, sizeof(int), (comparison_userdata_fn_t)cmp_int, NULL);`
> 
> But, there was a warning when I was trying to build the project for testing. the `comparison_userdata_fn_t` takes in three argument while the `cmp_int` and `cmpt_uint16` takes in two argument. But, there is actually a function `comparison_fn_t` which takes in two argument. I am thinking of changing the `comparison_userdata_fn_t` to `comparison_fn_t` . I also I could not find where the function x_bsearch was used. Kindly give advice on this. Thank you.

Thanks for your reply @bluca , I hope you are doing good ?  Also, kindly help look into this also. Thank you.

**Comment by Unique-Usman** [comment]
> > Hello @bluca, some question which I have regarding the integration test.
> > 
> > 1. I got to know that this link https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/control#L121 contains the depencies to run the integration test, my question is do we have to install of the dependency ?
> > 2. When I  checked the coverall, I saw some  percentage and some of them seems to be for unit test, how do I know which one is the unit test and which one is the integration test? Also, I do not really understand the structure in the coverall very well.
> > 
> > Thank you very much, I will be expecting your reply.
> 
> 1. Yes
> 2. That is not tracked, it's a single coverage value for everything

Thanks for your reply, how do I then know which integration has not been covered ? Thank you.

**Comment by devdekunle** [comment]
Hello @bluca, I hope you are doing fine, I has completed this task I asked if i coould do

> Hello @bluca, I found two tasks in the TODO list that I would be interested in.
> 
> 1. introduce an option (or replacement) for "systemctl show" that outputs all
>    properties as JSON, similar to busctl's new JSON output. In contrast to that
>    it should skip the variant type string though.
> 2. maybe add a tool that displays most recent journal logs as QR code to scan
>    off screen and run it automatically on boot failures, emergency logs and
>    such. Use DRM APIs directly, see
>    https://github.com/dvdhrm/docs/blob/master/drm-howto/modeset.c for an example
>    for doing that.
> 
> Please I would like to know your take on them with regards to existing or current implementations and maybe time expectancy for completing them. Though as regards implementation, I checked the systemctl show --output=json and It outputs key-value pairs, same as just systemctl show. Which i feel could mean it hasn't been implemented. So please can i get your take on the two?

Hello @bluca, I hope you are good. I had completed this task i asked if i could do and opened a PR #31948, unfortunately, I got to know implementation of --json option on the Dbus property had been stopped and generation of JSON should be done by the daemon instead. I wanted to take on the challenge of implementing it on PID1 but someone was already working on it.
I also want to ask if i could do this task i also found in the TODO page also.

* Add a "systemctl list-units --by-slice" mode or so, which rearranges the
  output of "systemctl list-units" slightly by showing the tree structure of
  the slices, and the units attached to them.

**Comment by devdekunle** [comment]
Hello sir @bluca. How are you doing. I hope you're good. 
Please sir I seek your review on this task I requested above #32020. Also I found two issues I'm interested in #30889 and #31359. Thank you sir and God bless you 🙏🏽.

**Comment by Unique-Usman** [comment]
> Hello @bluca, I hope you are good.
> 
> The deadline for final application is tommorrow,
> 
> 1. There is question asking about
>    `(Optional) Community-specific Questions`,
> 
> is there any one ?
> 
> 2. Also,there is another question
>    `Please work with your mentor to provide a timeline of the work you plan to accomplish on the project and what tasks you will finish at each step.`
> 
> What is your suggestion regarding this ?
> 
> 3. Also, is there any other information the community is expecting the applicant to fill in their application ?
> 4. I also opened a RFE PR [Implement the --json option for bootctl status #32035](https://github.com/systemd/systemd/pull/32035), kindly help review.
> 
> Thank you very much, I really so much appreciate your time.

Thank you @bluca for the review, kindly help answer the question regarding the final application which will be due in few hours.

**Comment by bluca** [comment]
> Hello @bluca, I hope you are good.
> 
> The deadline for final application is tommorrow,
> 
> 1. There is question asking about
>    `(Optional) Community-specific Questions`,
> 
> is there any one ?
> 
> 2. Also,there is another question
>    `Please work with your mentor to provide a timeline of the work you plan to accomplish on the project and what tasks you will finish at each step.`
> 
> What is your suggestion regarding this ?
> 
> 3. Also, is there any other information the community is expecting the applicant to fill in their application ?
> 4. I also opened a RFE PR [Implement the --json option for bootctl status #32035](https://github.com/systemd/systemd/pull/32035), kindly help review.
> 
> Thank you very much, I really so much appreciate your time.

1) no
2) just look at the project description and provide a rough estimate, it doesn't have to be precise
3) no
